File: backend/ai_engine/quality_checker.py
import math

class QualityChecker:
    def __init__(self):
        pass

    def check(self, landmarks, image_size):
        if not landmarks or len(landmarks) == 0:
            return {'fullBodyVisible': False, 'poseOk': False, 'lightingOk': False, 'messages': ["No landmarks detected"]}

        nose = self._find(landmarks, 0)
        l_shoulder = self._find(landmarks, 11)
        r_shoulder = self._find(landmarks, 12)
        l_hip = self._find(landmarks, 23)
        r_hip = self._find(landmarks, 24)
        l_ankle = self._find(landmarks, 27)
        r_ankle = self._find(landmarks, 28)

        missing = []
        if not self._is_visible(nose): missing.append("Head")
        if not self._is_visible(l_shoulder) or not self._is_visible(r_shoulder): missing.append("Shoulders")
        if not self._is_visible(l_hip) or not self._is_visible(r_hip): missing.append("Hips")
        if not self._is_visible(l_ankle) or not self._is_visible(r_ankle): missing.append("Feet")

        full_body_visible = len(missing) == 0

        pose_ok = True
        pose_messages = []

        if full_body_visible:
            shoulder_slope = abs((l_shoulder.y * image_size[1]) - (r_shoulder.y * image_size[1]))
            # 40 pixels was the heuristic, we assume image_size = (width, height) in pixels
            if shoulder_slope > 40:
                pose_ok = False
                pose_messages.append("Shoulders not level")
        else:
            pose_ok = False
            pose_messages.append("Body parts missing")

        lighting_ok = True

        essential_types = [11, 12, 23, 24]
        
        essential_visible = True
        for type_enum in essential_types:
            lm = self._find(landmarks, type_enum)
            if not lm or lm.visibility <= 0.5:
                essential_visible = False
                break

        messages = []
        if len(missing) > 0:
            messages.append(f"Vücudun bazı bölümleri (ör: {', '.join(missing)}) tam görünmüyor.")
        if not essential_visible:
            messages.append("Gövde net seçilemiyor, sonuçlar sadece tahminidir.")
        
        messages.extend(pose_messages)

        confidence = 1.0
        
        if len(missing) > 0:
            confidence -= 0.15 * len(missing)
        
        if not essential_visible:
            confidence -= 0.35

        avg_likelihood = sum([lm.visibility for lm in landmarks]) / len(landmarks)
        confidence *= avg_likelihood

        l_wrist = self._find(landmarks, 15)
        r_wrist = self._find(landmarks, 16)
        hands_in_pockets = False

        if l_wrist and r_wrist and l_hip and r_hip:
            # We must multiply by dimensions or just use normalized distance if both are normalized.
            # In Flutter, ML Kit gives absolute pixel coordinates. MediaPipe gives normalized [0.0, 1.0].
            # We must un-normalize for pixel distance.
            l_dist = self._dist(l_wrist, l_hip, image_size)
            r_dist = self._dist(r_wrist, r_hip, image_size)
            if l_dist < 0.1 * image_size[1] and r_dist < 0.1 * image_size[1]:
                hands_in_pockets = True
                confidence -= 0.1
                messages.append("Daha iyi analiz için ellerinizi görünür tutun.")

        if confidence < 0.05:
            confidence = 0.05

        return {
            'fullBodyVisible': full_body_visible,
            'essentialVisible': essential_visible,
            'poseOk': pose_messages == [], # True if empty
            'lightingOk': lighting_ok,
            'confidenceScore': confidence,
            'handsInPockets': hands_in_pockets,
            'messages': messages,
            'missingParts': missing,
        }

    def _dist(self, p1, p2, image_size):
        # MediaPipe returns normalized coordinates
        x1, y1 = p1.x * image_size[0], p1.y * image_size[1]
        x2, y2 = p2.x * image_size[0], p2.y * image_size[1]
        return math.sqrt((x1 - x2)**2 + (y1 - y2)**2)

    def _find(self, landmarks, landmark_type):
        try:
            return landmarks[landmark_type]
        except IndexError:
            return None

    def _is_visible(self, landmark):
        return landmark is not None and landmark.visibility > 0.5

File: flutter_app/lib/ai/engine/quality_checker.dart
import 'dart:ui';
import 'package:google_mlkit_pose_detection/google_mlkit_pose_detection.dart';
import 'dart:math';

class QualityChecker {
  Map<String, dynamic> check(List<PoseLandmark> landmarks, Size imageSize) {
    if (landmarks.isEmpty) {
      return {'fullBodyVisible': false, 'poseOk': false, 'lightingOk': false, 'messages': ["No landmarks detected"]};
    }

    // Map landmarks for easier access
    // ML Kit PoseLandmarkType enum:
    // 0: nose, 11: leftShoulder, 12: rightShoulder, 23: leftHip, 24: rightHip, 27: leftAnkle, 28: rightAnkle
    
    final nose = _find(landmarks, PoseLandmarkType.nose);
    final lShoulder = _find(landmarks, PoseLandmarkType.leftShoulder);
    final rShoulder = _find(landmarks, PoseLandmarkType.rightShoulder);
    final lHip = _find(landmarks, PoseLandmarkType.leftHip);
    final rHip = _find(landmarks, PoseLandmarkType.rightHip);
    final lAnkle = _find(landmarks, PoseLandmarkType.leftAnkle);
    final rAnkle = _find(landmarks, PoseLandmarkType.rightAnkle);

    // 1. Full Body Visibility
    // We need at least nose, shoulders, hips, and ankles to be somewhat visible
    bool fullBodyVisible = true;
    List<String> missing = [];
    
    if (!_isVisible(nose)) missing.add("Head");
    if (!_isVisible(lShoulder) || !_isVisible(rShoulder)) missing.add("Shoulders");
    if (!_isVisible(lHip) || !_isVisible(rHip)) missing.add("Hips");
    if (!_isVisible(lAnkle) || !_isVisible(rAnkle)) missing.add("Feet");

    if (missing.isNotEmpty) fullBodyVisible = false;

    // 2. Pose Ok (Upright)
    bool poseOk = true;
    List<String> poseMessages = [];

    if (fullBodyVisible) {
        // Shoulder slope
        double shoulderSlope = (lShoulder!.y - rShoulder!.y).abs();
        if (shoulderSlope > 40) { // Tolerant threshold in pixels, relative but ok for now
            poseOk = false;
            poseMessages.add("Shoulders not level");
        }
    } else {
        poseOk = false;
        poseMessages.add("Body parts missing");
    }

    // 3. Lighting (Simple placeholder)
    bool lightingOk = true; 

    // Essential landmarks for ANY analysis (Shoulders + Hips)
    List<PoseLandmarkType> essential = [
        PoseLandmarkType.leftShoulder, PoseLandmarkType.rightShoulder,
        PoseLandmarkType.leftHip, PoseLandmarkType.rightHip
    ];
    bool essentialVisible = essential.every((type) => landmarks.any((l) => l.type == type && l.likelihood > 0.5));

    List<String> messages = [];
    if (missing.isNotEmpty) {
        messages.add("Vücudun bazı bölümleri (ör: ${missing.join(', ')}) tam görünmüyor.");
    }
    if (!essentialVisible) {
        messages.add("Gövde net seçilemiyor, sonuçlar sadece tahminidir.");
    }
    messages.addAll(poseMessages);

    // 4. Confidence Calculation
    double confidence = 1.0;
    
    // Penalize missing parts but don't zero out
    if (missing.isNotEmpty) {
        confidence -= 0.15 * missing.length;
    }
    if (!essentialVisible) {
        confidence -= 0.35;
    }

    // Likelihood penalty
    double avgLikelihood = landmarks.map((e) => e.likelihood).reduce((a, b) => a + b) / landmarks.length;
    confidence *= avgLikelihood;

    // Hands in pockets heuristic (Wrist near hip)
    final lWrist = _find(landmarks, PoseLandmarkType.leftWrist);
    final rWrist = _find(landmarks, PoseLandmarkType.rightWrist);
    bool handsInPockets = false;
    if (lWrist != null && rWrist != null && lHip != null && rHip != null) {
        double lDist = _dist(lWrist, lHip);
        double rDist = _dist(rWrist, rHip);
        if (lDist < 0.1 * imageSize.height && rDist < 0.1 * imageSize.height) {
            handsInPockets = true;
            confidence -= 0.1;
            messages.add("Daha iyi analiz için ellerinizi görünür tutun.");
        }
    }

    if (confidence < 0.05) confidence = 0.05;

    return {
      'fullBodyVisible': missing.isEmpty,
      'essentialVisible': essentialVisible,
      'poseOk': poseMessages.isEmpty,
      'lightingOk': lightingOk,
      'confidenceScore': confidence,
      'handsInPockets': handsInPockets,
      'messages': messages,
      'missingParts': missing,
    };
  }

  double _dist(PoseLandmark p1, PoseLandmark p2) {
    return sqrt(pow(p1.x - p2.x, 2) + pow(p1.y - p2.y, 2));
  }

  PoseLandmark? _find(List<PoseLandmark> landmarks, PoseLandmarkType type) {
    try {
      return landmarks.firstWhere((element) => element.type == type);
    } catch (_) {
      return null;
    }
  }

  bool _isVisible(PoseLandmark? landmark) {
    return landmark != null && landmark.likelihood > 0.5;
  }
}
